### PR TITLE
gitchecker: add support for versions property

### LINKED ISSUE #291: Git checker doesn't understand version contraining
Even when restricting the checker to one version, f-e-d-c still suggests the actual latest tag.

```yaml
id: com.virustotal.Uploader
modules:
  - name: extra-cmake-modules
    sources:
      - type: git
        url: https://invent.kde.org/frameworks/extra-cmake-modules.git
        tag: v5.77.0
        commit: 760a013c3ff553b23d456c7787ed76981ec5142f
        x-checker-data:
          type: git
          tag-pattern: ^v(\d[\d.]+\d)$
          versions:
            ==: "1.10.1"
```

```
$ flatpak run org.flathub.flatpak-external-data-checker com.virustotal.Uploader.yaml
INFO    src.manifest: Checking 1 external data items
INFO    src.manifest: Started check [1/1] git extra-cmake-modules.git (from com.virustotal.Uploader.yaml)
INFO    src.lib.externaldata: Source extra-cmake-modules.git: got new version 5.93.0
INFO    src.manifest: Finished check [1/1] git extra-cmake-modules.git (from com.virustotal.Uploader.yaml)
OUTDATED: extra-cmake-modules.git
 Has a new version:
  URL:       https://invent.kde.org/frameworks/extra-cmake-modules.git
  Commit:    eccf695824fa931bca471bad3f1aa3a1f244fdd6
  Tag:       v5.93.0
  Branch:    None
  Version:   5.93.0
  Timestamp: None

INFO    src.main: Check finished with 0 error(s)
```

### PR DESCRIPTION
**Pull request by vchernin**
Fixes https://github.com/flathub/flatpak-external-data-checker/issues/291

### CODE REVIEW

File: src/checkers/gitchecker.py [hunk 1]
```diff
@@ -119,12 +122,23 @@ async def _check_has_new(external_data: ExternalGitRepo):
...
-                f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
-            ) from err
+        if constraints:
+            matching_versions = [tag.version for tag in matching_tags]
+            versions = filter_versions(matching_versions, constraints, sort=False)
+            try:
+                latest_version = versions[-1]
+                for found_tag in matching_tags:
+                    if found_tag.version is latest_version:
+                        latest_tag = found_tag
```
Comment (gasinvein) [inline src/checkers/gitchecker.py]: `filter_versions()` supports `to_string` arg. If you pass a callable to it, you can filter `TagWithVersion` objects directly, without converting them to version string list first.
E.g. `filter_versions(matching_tags, to_string = lambda t: t.version)` should work.
This way you won't need to iterate over `matching_tags` again, looking for a tag with matching version.

Comment (gasinvein) [reply src/checkers/gitchecker.py]: Overall, I'd change it like this:
```patch
diff --git a/src/checkers/gitchecker.py b/src/checkers/gitchecker.py
index f4d5bea..3d5db67 100644
--- a/src/checkers/gitchecker.py
+++ b/src/checkers/gitchecker.py
@@ -98,6 +98,7 @@ class GitChecker(Checker):
 
         version_scheme = external_data.checker_data.get("version-scheme", "loose")
         tag_cls = TAG_VERSION_SCHEMES[version_scheme]
+        sort_tags = external_data.checker_data.get("sort-tags", True)
         constraints = external_data.checker_data.get("versions", {}).items()
 
         matching_tags = []
@@ -117,28 +118,24 @@ class GitChecker(Checker):
             version = tag_match.group(1)
             matching_tags.append(tag_cls(commit, tag, annotated, version))
 
-        if external_data.checker_data.get("sort-tags", True):
+        if constraints:
+            sorted_tags = filter_versions(
+                matching_tags,
+                constraints=constraints,
+                to_string=lambda t: t.version,
+                sort=sort_tags,
+            )
+        elif sort_tags:
             sorted_tags = sorted(matching_tags)
         else:
             sorted_tags = matching_tags
 
-        if constraints:
-            matching_versions = [tag.version for tag in matching_tags]
-            versions = filter_versions(matching_versions, constraints, sort=False)
-            try:
-                latest_version = versions[-1]
-                for found_tag in matching_tags:
-                    if found_tag.version is latest_version:
-                        latest_tag = found_tag
-            except IndexError as err:
-                raise CheckerQueryError("Can't find matching version") from err
-        else:
-            try:
-                latest_tag = sorted_tags[-1]
-            except IndexError as err:
-                raise CheckerQueryError(
-                    f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
-                ) from err
+        try:
+            latest_tag = sorted_tags[-1]
+        except IndexError as err:
+            raise CheckerQueryError(
+                f"{external_data.current_version.url} has no tags matching '{tag_pattern}'"
+            ) from err
 
         new_version = ExternalGitRef(
             url=external_data.current_version.url,

```

Comment (vchernin) [reply src/checkers/gitchecker.py]: Thanks, that is much cleaner.

**Comment by gasinvein** [comment]
Looks good, can you please also add a note about version constraining to the readme, similar to this?
https://github.com/flathub/flatpak-external-data-checker/blob/dfac7bc79493a9380c34a134b1eaf05c6eefece4/README.md?plain=1#L308

**Comment by vchernin** [comment]
> Looks good, can you please also add a note about version constraining to the readme, similar to this?
> 
> https://github.com/flathub/flatpak-external-data-checker/blob/dfac7bc79493a9380c34a134b1eaf05c6eefece4/README.md?plain=1#L308

Done

**Comment by gasinvein** [comment]
Thanks!

**Pull request merged**